### DISCUSSION TITLE
Auto index: poor performance on `exists` queries

### DISCUSSION THREAD

**Discussion by danielwinkler** [ravendb/ravendb#18698] [Q&A] [answered]
Dear community,

We have a collection of `Assignments` that we want to filter with the query

```
from 'Assignments'
where StudentId = "students-11111-A"
and exists(Trainings)
```

- If we remove the `exists(Trainings)` query, the result is returned 1 ms
- If we add the query, the time goes up to 500ms - 1000ms

using `include timings()` I can see that the time is spent on Lucene

The used auto index is indexing the `Trainings` field, which can be seen from the definition as well as from the name
Auto/Assignments/By...StudentIdAndTrainings

- Is this an expected behavior?
- If so, would the alternative be to create a static index that computes the TrainingCount with a 0 fallback and verify it's > 0?

Thanks in advance,
daniel

**Comment by ayende** [comment, marked as answer]
I assume that most documents have a `Trainings` field? 

Also note that I assume that this is an array? So you are actually indexing the _JSON_ of the contents of the array.
That will likely create a lot of individual values for this, and we have to scan through all of them.

What if you do:

```
from 'Assignments'
where StudentId = "students-11111-A"
and Trainings.Length > 0
```

Note that empty array _will_ exist.

**Comment by danielwinkler** [comment]
> I assume that most documents have a Trainings field?

Correct

> What if you do: ...

```
from 'Assignments'
where StudentId = "students-11111-A"
and Trainings.Length > 0
```
The query re-indexed the auto-index and added `Trainings.Length`, after that it is returning in 2ms

It also filters the non-existing fields correctly, so the query would be perfect.

> Note that empty array will exist.

Is this behavior somewhere documented? Because for primitive fields i need to make an exists query, a null check won't work.

**Reply by ayende** [reply]
The issue is specifically around a problem with indexing arrays (or objects), which generate tremendous number of unique values.
That is actually an error in how this work, but we can't easily fix that with Lucene because of backward compact concerns.

Corax will reject this query.

**Reply by danielwinkler** [reply]
I see, thank you.

So if we change everything to Corax, we will get an error that indicate to query for the non-zero length of arrays?

**Reply by ayende** [reply]
Afraid not, it will just tell you that you cannot index arrays.

The issue is that it never makes sense to index an _array_, because you aren't searching for it.


Another aspect is that `exists` itself is a bit annoying. Because it test for the _presence_ of a field, not it's value. That make things a bit harder for the query engine when you have a _lot_ of unique values

**Reply by danielwinkler** [reply]
Maybe it would be good idea to add the information about arrays / complex objects here
https://ravendb.net/docs/article-page/6.0/csharp/client-api/session/querying/how-to-filter-by-field

Or consider the type of the field, and rewrite the exists(Trainings) query to a Length > 0 query internally.
But there are probably good reasons why this is not so easy (e.g. exists("Trainings.IsActive"))

And as last question, if we would have a complex object `Stats` on `Training`.
Would there also be an efficient way to check the presence, I assume `exists` will again scan through all serialized objects?

**Reply by ayende** [reply]
For the complex object, I would probably suggest just using a static index and adding the `StatsExists` field there.
